recipe/meta.yaml

```diff
@@ -10,6 +10,8 @@
 {% endif %}
 
 {% set mkl = "<2024" %}
+# see .ci/docker/ci_commit_pins/triton.txt
+{% set triton = "3.1.0" %}
```

Comment:
How much does triton change? is it impossible for us to release this pin?

Basically, by pinning so hard at runtime, we will need to "maintain triton 3.1.0" which can lead to instability when users install a package with a different triton requirement

Reply:
To be honest, I don't know but they seem to be released in tandem:

https://github.com/pytorch/pytorch/blob/main/RELEASE.md#triton-dependency-for-the-release

Upstream wheels also pin to the very specific `triton (==3.1.0)` version (but then, they pin CUDA to very specific versions as well).

Reply:
hmm ok, interesting
```suggestion
# pytorch and triton are released in tandem, see notes in their release process
# https://github.com/pytorch/pytorch/blob/main/RELEASE.md#triton-dependency-for-the-release
{% set triton = "3.1.0" %}
```

lets not push this right now, but I think it will be good to "document" what you said.

Reply:
Will do, added it locally now, for when the time for the next push comes.

Reply:
thanks marking this as resolved!

Reply:
For completeness: yes, the `triton` pin is critical and needs to be exact. Every Triton upgrade is a lot of work, supporting a version range is infeasible. It's much more fragile than say CUDA or MKL.

Reply:
hmmm.. take this as a user story:

For some reason, our crazy pipeline has both tensorflow and pytorch. One day we will fix it (trust me we are trying).
I would much rather have:
* Medium performance pytorch + gpu, and tensorflow + gpu

rather than
* pytorch + cpu and tensorflow +gpu

Often, due to versions conflicts accross conda forge, see for example https://github.com/conda-forge/tensorflow-feedstock/pull/407 mamba and conda get confused about what to pick. I'm generally worried that `triton` may make pytroch-gpu less compatible with other hard to compile gpu packages (such as tensorflow).

Could we have a "slower rollout" of `triton + gpu` and add it as a "run_constrained" instead? In 6-12 months, we could re-assess provided many here are still involved?

Generally speaking, without much of your (renewed) involvement, I was about to consider pytorch/tensorflow on lifesupport....

Reply:
TensorFlow doesn't depend on Triton though? Triton is a hard dependency and `torch.compile` (the flagship PyTorch 2.0 feature) on GPU depends on it, so I wouldn't make it optional unless there's a very compelling reason (and then only temporary so that the actual problem can be fixed).

Is there an actual problem resolving PyTorch + TF after this change? (I can't test this week, traveling, sorry)

Reply:
> Is there an actual problem resolving PyTorch + TF after this change? (I can't test this week, traveling, sorry)

not "after this change" but it does add many extra runtime dependencies which may cause the problem to get worse.


> Triton is a hard dependency and torch.compile

The thing is that the value proposition of `torch.compile` is really only valuable in specific scenerios and those are still an area of active ~research~ exploration.
http://blog.ezyang.com/2024/11/ways-to-use-torch-compile/

>  torch.compile (the flagship PyTorch 2.0 feature)

I'm not sure that JITs are really all they are cracked up to be. They lead to alot of "late stage failures" and we try to avoid that whenever we can my work. conda-forge has helped ALOT to ensure llvm versions match at runtime, but you know, they still pop up. We still hit "unable to connect to GPU errors" when the underlying driver gets updated by ubuntu.....

The extracted size of the environment increase from 6.1GB to 7.4GB when installing triton ontop of 2.5.1

<details><summary>triton env creation</summary>

```
$ conda create --name pt "pytorch=2.5.1=cuda*" python=3.12 --channel conda-forge --override-channels
$ conda create --name pt_triton "pytorch=2.5.1=cuda*" python=3.12 triton --channel conda-forge --override-channels

$ du -sh pt/ && du -sh pt_triton/
6.1G	pt/
7.4G	pt_triton/
```
</details>
<details><summary>incompatibility is due to libabseil</summary>

Yea today, tensorflow 2.17 and the latest pytorch 2.5.1 builds are incompatible due to https://github.com/conda-forge/tensorflow-feedstock/pull/407


```
$ conda create --name tf_pytorch "pytorch=2.5.1=*_7" "tensorflow=2.17.0=*3" --channel conda-forge --override-channels
Channels:
 - conda-forge
Platform: linux-64
Collecting package metadata (repodata.json): done
Solving environment: failed

LibMambaUnsatisfiableError: Encountered problems while solving:
  - nothing provides __cuda needed by tensorflow-2.17.0-cuda120py310heb3ae67_203

Could not solve for environment specs
The following packages are incompatible
├─ pytorch 2.5.1 *_7 is installable and it requires
│  └─ libprotobuf >=5.28.2,<5.28.3.0a0 , which can be installed;
└─ tensorflow 2.17.0 *3 is not installable because there are no viable options
   ├─ tensorflow 2.17.0 would require
   │  └─ tensorflow-base [2.17.0 cpu_py310hfda4fce_3|2.17.0 cpu_py311h78ee05f_3|2.17.0 cpu_py312h4c54302_3|2.17.0 cpu_py39h9b712fd_3], which requires
   │     └─ libgrpc >=1.65.5,<1.66.0a0 , which requires
   │        └─ libprotobuf >=5.27.5,<5.27.6.0a0 , which conflicts with any installable versions previously reported;
   └─ tensorflow 2.17.0 would require
      └─ __cuda, which is missing on the system.
```


allowing for tensorflow as old as 2.10
```
$ conda create --name tf_pytorch "pytorch=2.5.1=*_7" "tensorflow>=2.10" --channel conda-forge --override-channels
Channels:
 - conda-forge
Platform: linux-64
Collecting package metadata (repodata.json): done
Solving environment: failed

LibMambaUnsatisfiableError: Encountered problems while solving:
  - package pytorch-2.5.1-cpu_generic_py310_h5d8512c_7 requires libprotobuf >=5.28.2,<5.28.3.0a0, but none of the providers can be installed

Could not solve for environment specs
The following packages are incompatible
├─ pytorch 2.5.1 *_7 is installable and it requires
│  └─ libprotobuf >=5.28.2,<5.28.3.0a0 , which can be installed;
└─ tensorflow >=2.10  is not installable because there are no viable options
   ├─ tensorflow 2.10.0 would require
   │  └─ tensorflow-base [2.10.0 cpu_py310hc537a0e_0|2.10.0 cpu_py37h50bd216_0|2.10.0 cpu_py38h67fe383_0|2.10.0 cpu_py39h16601f7_0], which requires
   │     └─ libprotobuf >=3.21.6,<3.22.0a0 , which conflicts with any installable versions previously reported;
   ├─ tensorflow [2.10.0|2.11.0|...|2.17.0] would require
   │  └─ __cuda, which is missing on the system;
   ├─ tensorflow 2.11.0 would require
   │  └─ tensorflow-base [2.11.0 cpu_py310*_0|2.11.0 cpu_py38*_0|2.11.0 cpu_py39*_0], which requires
   │     └─ libprotobuf >=3.21.11,<3.22.0a0 , which conflicts with any installable versions previously reported;
   ├─ tensorflow [2.11.0|2.11.1|...|2.15.0] would require
   │  └─ tensorflow-base [2.11.0 cpu_py310h9fe5a9a_1|2.11.0 cpu_py38h1497e92_1|...|2.15.0 cpu_py39h65ed569_0], which requires
   │     └─ libprotobuf >=3.21.12,<3.22.0a0 , which conflicts with any installable versions previously reported;
   ├─ tensorflow 2.15.0 would require
   │  └─ tensorflow-base [2.15.0 cpu_py310h7e4d085_1|2.15.0 cpu_py310h7e4d085_2|...|2.15.0 cpu_py39he1df281_2], which requires
   │     └─ libprotobuf >=4.24.4,<4.24.5.0a0 , which conflicts with any installable versions previously reported;
   ├─ tensorflow [2.16.1|2.16.2|2.17.0] would require
   │  └─ tensorflow-base [2.16.1 cpu_py310h224022f_0|2.16.1 cpu_py311h2e85f28_0|...|2.17.0 cpu_py39h6ebc309_0], which requires
   │     └─ libprotobuf >=4.25.3,<4.25.4.0a0 , which conflicts with any installable versions previously reported;
   └─ tensorflow 2.17.0 would require
      └─ tensorflow-base [2.17.0 cpu_py310hfda4fce_3|2.17.0 cpu_py311h78ee05f_3|2.17.0 cpu_py312h4c54302_3|2.17.0 cpu_py39h9b712fd_3], which requires
         └─ libprotobuf >=5.27.5,<5.27.6.0a0 , which conflicts with any installable versions previously reported.
```

</details>

Reply:
I don't think it's an issue to have an exact pin on triton - it just means we build them in sync, but that's manageable. Triton only has one other consumer currently - running
```
mamba repoquery whoneeds -c conda-forge --platform linux-64 triton
```
shows only `pyg-lib` (which in turn is used only by `pytorch_geometric`). IOW, there's nothing to conflict with, except packages that are already in the pytorch universe.

Looking at the [dependencies](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Flinux-64%2Ftriton-3.1.0-cuda126py39h6b1a799_2.conda) of triton, there's not much that stands out (and we'll be adding `cuda-cupti` through #305 or similar anyway).

I think we should take the same approach as for the magma debate: if it's supported by default upstream, match it in the feedstock. Finally,

The abseil issues are IMO completely independent of the triton debate. Basically, the TF feedstock would need a lot more love (we've been stuck on the recent abseil [migration](https://github.com/conda-forge/tensorflow-feedstock/pull/407) and 2.18 for >2 months; which would also be necessary for numpy 2.0), not helped by the fact that TF is hairier to update w.r.t. abseil/protobuf/etc. than pytorch. That won't change based on whether we add triton here or not.

Reply:
> The extracted size of the environment increase from 6.1GB to 7.4GB when installing triton ontop of 2.5.1

That does look like it needs a careful look, because that's a large increase (not impossible there's another compression issue like with `libmagma` for example). I agree with @h-vetinari that not much jumps out immediately, aside from `pytorch * cuda*` that needs removing. The rest do seem to be dependencies that are undeclared in the triton wheels but do actually get installed when building `triton` from source (ironically, from a conda channel):

https://github.com/triton-lang/triton/blob/4c91b8154a5061c47207ce8c04ee6ac98b8231b9/python/setup.py#L533-L541



> The thing is that the value proposition of `torch.compile` is really only valuable in specific scenerios and those are still an area of active ~research~ exploration.
> http://blog.ezyang.com/2024/11/ways-to-use-torch-compile/

Well that blog shows it's not a magic bullet, but also that a lot of production pipelines already depend on it, as well as downstream tools like `vllm`. It is a core feature that is pretty widely used, so making it opt-in feels wrong.

Reply:
Alright. Thanks for the discussion

Reply:
I’m fine if the strategy is let’s get to feature parity with upstream pypi